python/pywatchman/aioclient.py

```diff
@@ -0,0 +1,340 @@
+        return bser.loads(response) # Defaults to BSER v1
+
+
+class ReceiveLoopError(Exception):
+    pass
+
+
+class AIOClient(object):
+
+    def __init__(self, connection):
+        self.connection   = connection
+        self.log_queue    = asyncio.Queue()
+        self.sub_by_root  = {}
+        self.bilateral_response_queue = asyncio.Queue()
+        self.receive_task = None
+
+    def stop(self):
+        self.should_stop = True
+
+    async def receive_bilateral_response(self):
+        '''Receive the response to a request made to the Watchman service.'''
+
+        self._check_recieve_loop()
+        resp = await self.bilateral_response_queue.get()
+        self._check_error(resp)
+        return resp
+
+    async def query(self, *args):
+        '''Send a query to the Watchman service and return the response.'''
+
+        self._check_recieve_loop()
+        try:
+            await self.connection.send(args)
+            return await self.receive_bilateral_response()
+        except CommandError as ex:
+            ex.setCommand(args)
+            raise ex
+
+    async def capability_check(self, optional=None, required=None):
+        '''Perform a server capability check.'''
+
+        self._check_recieve_loop()
+        # If the returned response is an error, self.query will raise an error
+        await self.query('version', {
+            'optional': optional or [],
+            'required': required or []
+        })
+
+    async def get_subscription(self, name, root):
+        ''' Retrieve the data associated with a named subscription
+
+        Returns None if there is no data associated with `name`
+
+        If root is not None, then only return the subscription
+        data that matches both root and name.  When used in this way,
+        remove processing impacts both the unscoped and scoped stores
+        for the subscription data.
+        '''
+        self._check_recieve_loop()
+        # People may pass in Unicode strings here -- but currently BSER only
+        # returns bytestrings. Deal with that.
+        # TODO: This won't be a problem with BSER v2
+        if isinstance(root, str):
+            root = encoding.encode_local(root)
+            root = os.path.normcase(root)
+        if isinstance(name, str):
+            name = name.encode('utf-8')
+        self._ensure_subscription_queue_exists(name, root)
+        return await self.sub_by_root[root][name].get()
+
+    async def pop_log(self):
+        '''Get one log from the log queue.'''
+        self._check_recieve_loop()
+        return await self.log_queue.get()
+
+    def close(self):
+        '''Close the underlying connection.'''
+        if self.receive_task:
+            self.receive_task.cancel()
+        if self.connection:
+            self.connection.close()
+
+    def enable_receiving(self, loop=None):
+        '''Schedules the receive loop to run on the given loop.'''
+        self.receive_task = asyncio.ensure_future(self._receive_loop(),
```

Comment:
You should schedule a callback for an exception happening in this task. It's an endless while loop that can't be broken if you cancel it (that's expected) or if you get any response that raises WatchmanError (that's unlikely but we should support recovery here). asyncio will be unhappy with you if there was an exception stored in the task and it was never read by your program.

Also, this doesn't seem like a tested path in the client's tests.